pkg/ingester/ingester_v2.go

```diff
@@ -1280,3 +1312,435 @@ func (i *Ingester) v2FlushHandler(w http.ResponseWriter, _ *http.Request) {
+
+	userDB := &userTSDB{
+		userID:              userID,
+		refCache:            cortex_tsdb.NewRefCache(),
+		ingestedAPISamples:  newEWMARate(0.2, i.cfg.RateUpdatePeriod),
+		ingestedRuleSamples: newEWMARate(0.2, i.cfg.RateUpdatePeriod),
+		lastUpdate:          atomic.NewInt64(0),
+	}
+
+	// Create a new user database
+	db, err := tsdb.Open(dbDir, userLogger, tsdbPromReg, &tsdb.Options{
+		RetentionDuration: i.cfg.TSDBConfig.Retention.Milliseconds(),
+		MinBlockDuration:  blockRange,
+		MaxBlockDuration:  blockRange,
+		NoLockfile:        true,
+		StripeSize:        i.cfg.TSDBConfig.StripeSize,
+		WALCompression:    i.cfg.TSDBConfig.WALCompressionEnabled,
+	})
+	if err != nil {
+		return nil, err
+	}
+	db.DisableCompactions() // we will compact on our own schedule
+
+	userDB.DB = db
+	userDB.setLastUpdate(time.Now()) // After WAL replay.
+
+	// Thanos shipper requires at least 1 external label to be set. For this reason,
+	// we set the tenant ID as external label and we'll filter it out when reading
+	// the series from the storage.
+	l := labels.Labels{
+		{
+			Name:  cortex_tsdb.TenantIDExternalLabel,
+			Value: userID,
+		}, {
+			Name:  cortex_tsdb.IngesterIDExternalLabel,
+			Value: i.lifecycler.ID,
+		},
+	}
+
+	// Create a new shipper for this database
+	if i.cfg.TSDBConfig.ShipInterval > 0 {
+		userDB.shipper = shipper.New(
+			userLogger,
+			tsdbPromReg,
+			udir,
+			cortex_tsdb.NewUserBucketClient(userID, i.TSDBState.bucket),
+			func() labels.Labels { return l },
+			metadata.ReceiveSource,
+			true, // Allow out of order uploads. It's fine in Cortex's context.
+		)
+	}
+
+	//i.TSDBState.tsdbMetrics.setRegistryForUser(userID, tsdbPromReg)
+	return userDB, nil
+}
+
+func (i *Ingester) v2BackfillPush(userID string, la []client.LabelAdapter, s client.Sample) error {
+	bucket, err := i.getOrCreateBackfillTSDB(userID, s.TimestampMs)
+	if err != nil {
+		return err
+	}
+
+	startAppend := time.Now()
+	db := bucket.db
+	bucket.lastAppend = startAppend
+	cachedRef, cachedRefExists := db.refCache.Ref(startAppend, client.FromLabelAdaptersToLabels(la))
+
+	app := db.Appender()
+	// If the cached reference exists, we try to use it.
+	if cachedRefExists {
+		err = app.AddFast(cachedRef, s.TimestampMs, s.Value)
+		if err != nil && errors.Cause(err) == storage.ErrNotFound {
+			cachedRefExists = false
+			err = nil
+		}
+	}
+
+	// If the cached reference doesn't exist, we (re)try without using the reference.
+	if !cachedRefExists {
+		var ref uint64
+
+		// Copy the label set because both TSDB and the cache may retain it.
+		copiedLabels := client.FromLabelAdaptersToLabelsWithCopy(la)
+
+		if ref, err = app.Add(copiedLabels, s.TimestampMs, s.Value); err == nil {
+			db.refCache.SetRef(startAppend, copiedLabels, ref)
+			cachedRef = ref
+			cachedRefExists = true
+		}
+	}
+
+	if err == nil {
+		err = app.Commit()
+	}
+
+	return err
+}
+
+// Assumes 1h bucket range for . TODO(codesome): protect stuff with locks.
+type backfillTSDBs struct {
```

Comment:
This `ingester_v2.go` file is getting quite big. I would move `backfillTSDBs` to a dedicated `.go` file.